Source/WebCore/html/ValidatedFormListedElement.cpp

```diff
@@ -0,0 +1,494 @@
+#include "HTMLParserIdioms.h"
+#include "PseudoClassChangeInvalidation.h"
+#include "ScriptDisallowedScope.h"
+#include "ValidationMessage.h"
+#include <wtf/Ref.h>
+#include <wtf/SetForScope.h>
+#include <wtf/Vector.h>
+
+namespace WebCore {
+
+using namespace HTMLNames;
+
+ValidatedFormListedElement::ValidatedFormListedElement(HTMLFormElement* form)
+    : FormListedElement(form)
+{
+}
+
+ValidatedFormListedElement::~ValidatedFormListedElement()
+{
+    clearForm();
+}
+
+bool ValidatedFormListedElement::willValidate() const
+{
+    if (!m_willValidateInitialized || m_dataListAncestorState == DataListAncestorState::Unknown) {
+        m_willValidateInitialized = true;
+        bool newWillValidate = computeWillValidate();
+        if (m_willValidate != newWillValidate)
+            m_willValidate = newWillValidate;
+    } else {
+        // If the following assertion fails, updateWillValidateAndValidity() is not
+        // called correctly when something which changes computeWillValidate() result
+        // is updated.
+        ASSERT(m_willValidate == computeWillValidate());
+    }
+    return m_willValidate;
+}
+
+bool ValidatedFormListedElement::computeWillValidate() const
+{
+    if (m_dataListAncestorState == DataListAncestorState::Unknown) {
+#if ENABLE(DATALIST_ELEMENT)
+        const HTMLElement& element = asHTMLElement();
+        m_dataListAncestorState = (element.document().hasDataListElements() && ancestorsOfType<HTMLDataListElement>(element).first()) ? DataListAncestorState::InsideDataList : DataListAncestorState::NotInsideDataList;
+#else
+        m_dataListAncestorState = DataListAncestorState::NotInsideDataList;
+#endif
+    }
+    // readonly bars constraint validation for *all* <input> elements, regardless of the <input> type, for compat reasons.
+    return m_dataListAncestorState == DataListAncestorState::NotInsideDataList && !isDisabled() && !m_hasReadOnlyAttribute;
+}
+
+void ValidatedFormListedElement::updateVisibleValidationMessage(RefPtr<HTMLElement> anchor)
+{
+    HTMLElement& element = asHTMLElement();
+    Page* page = element.document().page();
+    if (!page || !anchor)
+        return;
+    String message;
+    if (element.renderer() && willValidate())
+        message = validationMessage().stripWhiteSpace();
+    if (!m_validationMessage)
+        m_validationMessage = makeUnique<ValidationMessage>(*anchor);
+    m_validationMessage->updateValidationMessage(*anchor, message);
+}
+
+void ValidatedFormListedElement::hideVisibleValidationMessage()
+{
+    if (m_validationMessage)
+        m_validationMessage->requestToHideMessage();
+}
+
+bool ValidatedFormListedElement::checkValidity(Vector<RefPtr<ValidatedFormListedElement>>* unhandledInvalidControls)
+{
+    if (!willValidate() || isValidFormControlElement())
+        return true;
+    // An event handler can deref this object.
+    HTMLElement& element = asHTMLElement();
+    Ref<ValidatedFormListedElement> protectedThis(*this);
```

Comment:
It's slightly ever more efficient to protect this as HTMLElement since that'll avoid virtual function call.

Reply:
Smart, fixed in [1513fd1](https://github.com/WebKit/WebKit/pull/4988/commits/1513fd15243340510ecb0040cdc555db85f76afe).